### PR TITLE
[박민상] 문자열 덧셈 계산기 및 로또 (1단계)

### PR DESCRIPTION
**Pull request by minsang0850**

### CODE REVIEW

File: src/test/java/calculator/AdderTest.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+        );
+
+        Adder adder = new Adder(tokens);
+        int result = adder.calculate();
+
+        assertThat(6).isEqualTo(result);
+    }
+
+    @Test
+    void 숫자_하나(){
```
Comment (hyunniiii) [inline src/test/java/calculator/AdderTest.java:28]: 테스트 네이밍은 `숫자_하나` 보다는 `숫자 하나를 입력했을 때 입력된 숫자만 반환한다.`로 작성하면 해당 코드에 대해 이해하기 조금 더 좋지 않을까요?

File: src/test/java/calculator/ParserTest.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    @Test
+    void 기본구분자_테스트() {
+        Parser parser = new Parser("1,2:3");
+        assertThat(parser.getDelimiter()).isEqualTo(",|:");
+    }
+
+    @Test
+    void 커스텀구분자_테스트() {
+        Parser parser = new Parser("//;\n1;2;3");
+        //String customDelimiter=parser.findDelimiter();
```
Comment (hyunniiii) [inline src/test/java/calculator/ParserTest.java:22]: 필요없는 주석은 삭제해주세요 ″̮

File: src/test/java/calculator/ParserTest.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    @ParameterizedTest
+    @NullAndEmptySource
+    void emptyOrNull(final String text) {
+        Parser parser = new Parser (text);
+        assertThat(parser.getUserInput()).isEqualTo("0");
+    }
+
+    @Test
+    void 숫자_하나(){
+        Parser parser = new Parser("1");
```
Comment (hyunniiii) [inline src/test/java/calculator/ParserTest.java:36]: `AdderTest`에 존재해서 따로 작성해주시지 않은걸까요?
사용하지 않는 코드는 삭제해주는건 어떨까요?

File: src/main/java/calculator/Parser.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
+package calculator;
+
+import java.util.List;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class Parser {
+
+    private String userInput;
```
Comment (hyunniiii) [inline src/main/java/calculator/Parser.java:9]: `Parser`가 상태를 갖지 않도록 하는건 어떨까요?
상태를 가지고 있지 않고 값을 메서드의 파라미터로 받아도 될 것 같습니다.

File: src/main/java/lotto/LottoAuto.java [hunk 5]
```diff
@@ -0,0 +1,15 @@
+package lotto;
+
+public class LottoAuto {
+
+    public void play() {
+        LottoRules lottoRules = new LottoRules();
+        User user = new User(
+          OutputView.getLottoCount(InputView.getPrice() / lottoRules.getLottoPrice()), lottoRules);
```
Comment (hyunniiii) [inline src/main/java/lotto/LottoAuto.java:8]: 가독성을 위해 변수로 받아주는건 어떨까요?

File: src/main/java/lotto/LottoRules.java [hunk 6]
```diff
@@ -0,0 +1,41 @@
+package lotto;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class LottoRules {
+
+    private final int START_NUM = 1;
```
Comment (hyunniiii) [inline src/main/java/lotto/LottoRules.java:8]: 자바에서 상수는 `static`키워드를 사용합니다. ″̮

File: src/main/java/lotto/LottoRules.java [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    private final int LAST_NUM = 45;
+    private final int LOTTO_COUNT = 6;
+    private final int LOTTO_PRICE = 1000;
+    private final List<Integer> LOTTO_NUMS;
```
Comment (hyunniiii) [inline src/main/java/lotto/LottoRules.java:12]: `LOTTO_NUMS`는 상수일까요 변수일까요?
두가지에 대한 규칙이 혼동되어 있는 것 같습니다.
https://kephilab.tistory.com/51

File: src/main/java/lotto/LottoRules.java [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+            lottoNums.add(i);
+        }
+        this.LOTTO_NUMS = lottoNums;
+    }
+
+    public List<Integer> getLottoNums() {
+        return LOTTO_NUMS;
+    }
+
+    public int getLottoPrice() {
```
Comment (hyunniiii) [inline src/main/java/lotto/LottoRules.java:26]: 상수를 외부에서 사용하기 위해 getter를 생성해주신거라면 private이 아닌 public으로 선언해주는건 어떨까요?
상수는 한번 초기화되면 수정이 불가하여 `public`으로 선언해주셔도 괜찮을 것 같습니다.
`LottoRules.LOTTO_PRICE`로 사용하실 수 있습니다.

File: src/main/java/lotto/OutputView.java [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+
+    public static void printLottos(List<Lotto> lottos) {
+        lottos.forEach(lotto -> System.out.println(lotto.getNumbers()));
+    }
+
+    public static void printResult(Map<Rank, Integer> ranks, int price) {
+        String loss = "";
+        System.out.println(WINNING_STATISTICS_MESSAGE);
+        double reward = printRanks(ranks);
+        double yield = reward / price;
```
Comment (hyunniiii) [inline src/main/java/lotto/OutputView.java:28]: `reward / price`는 비즈니스 로직인 것 같습니다.
뷰에는 비즈니스 로직이 들어가지 않는것이 좋습니다 ″̮ 
해당 로직에 대한 책임을 가지고 있는 곳으로 옮겨주는건 어떨까요?

File: src/main/java/lotto/Parser.java [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+public class Parser {
+
+    public static List<Integer> parseWinningNumber(String numbersString) {
+        numbersString = numbersString.replace(" ", "");
+        String[] numberStringList = numbersString.split(",");
+        int listSize = numberStringList.length;
+        List<Integer> numbers = new ArrayList<>();
+        for (int i = 0; i < listSize; i++) {
+            numbers.add(Integer.parseInt(numberStringList[i]));
+        }
```
Comment (hyunniiii) [inline src/main/java/lotto/Parser.java:15]:
```suggestion
        for (String s : numberStringList) {
            numbers.add(Integer.parseInt(s));
        }
```

이렇게도 표현이 가능합니다.

File: src/test/java/lotto/RankTest.java [hunk 11]
```diff
@@ -0,0 +1,14 @@
+package lotto;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import org.junit.jupiter.api.Test;
+
+class RankTest {
+
+    @Test
+    void 랭크를_잘_받는지_확인() {
```
Comment (hyunniiii) [inline src/test/java/lotto/RankTest.java:10]: 테스트할 수 있는 케이스가 더 많아보입니다 ″̮

File: src/main/java/lotto/WinningNumber.java [hunk 12]
```diff
@@ -0,0 +1,28 @@
+package lotto;
+
+import java.util.List;
+
+public class WinningNumber {
+
+    private final List<Integer> winningNumber;
```
Comment (hyunniiii) [inline src/main/java/lotto/WinningNumber.java:7]: `Lotto` 클래스를 이용해보는건 어떨까요?

**Comment by hyunniiii** [comment]
도메인, 뷰, 컨트롤러 패키지를 생성해서 분리해주세요 ″̮

**Review by hyunniiii — CHANGES_REQUESTED** [review]
안녕하세요 민상님~
몇가지 코멘트 남겼습니다!
확인부탁드립니다 😄

File: src/main/java/lotto/WinningNumber.java (hunk 12 above)
Comment (minsang0850) [reply src/main/java/lotto/WinningNumber.java:7]: private final List<Integer> winningNumber;
에 privagte final Lotto 를 쓰라는 말씀이신가요?

저는 로뚜 규칙에 맞는 로또 번호를 Lotto의 생성자에서 아예 생성해줬는데
winningNumber는 임의로 받는 입력이라 어떻게 합쳐야할지 잘 모르겠습니다!

File: src/main/java/lotto/Parser.java [hunk 13]
```diff
@@ -0,0 +1,18 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class Parser {
+
+    public static List<Integer> parseWinningNumber(String numbersString) {
+        numbersString = numbersString.replace(" ", "");
+        String[] numberStringList = numbersString.split(",");
+        int listSize = numberStringList.length;
```
Comment (hyunniiii) [inline src/main/java/lotto/Parser.java:11]: `listSize`는 사용하고 있지 않네요.
지워주는건 어떨까요?

File: src/main/java/lotto/Statistics.java [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+    public static void setRanks(WinningNumber winningNumber, List<Lotto> lottos) {
+        for (Lotto lotto : lottos) {
+            int matchCount = getMatchCount(lotto.getNumbers(),
+              winningNumber.getWinningNumber());
+            boolean bonusScore = getBonusScore(lotto.getNumbers(),
+              winningNumber.getBonusBall());
+            Rank rank = getRank(matchCount, bonusScore);
+            incrementRank(rank);
+        }
+    }
```
Comment (hyunniiii) [inline src/main/java/lotto/Statistics.java:29]: 아무것도 반환하지 않는 함수들은 참조를 넘겨받아 참조에 직접 작업을 수행하는 사이드 이펙트를 유발하는 함수입니다.

참조를 넘겨 메소드를 호출하기보다 명확한 반환값을 넣고 반환 값을 받아 코드를 수행하도록 하면 어떨까요?

File: src/main/java/lotto/WinningNumberValidator.java [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+        validateRange(integer);
+    }
+    
+    private void validateRange(Integer integer) {
+        if (integer < lottoRules.START_NUM) {
+            throw new RuntimeException(INCORRECT_INPUT);
+        }
+        if (integer > lottoRules.LAST_NUM) {
+            throw new RuntimeException(INCORRECT_INPUT);
+        }
```
Comment (hyunniiii) [inline src/main/java/lotto/WinningNumberValidator.java:37]:
```suggestion
        if (integer < LottoRules.START_NUM) {
            throw new RuntimeException(INCORRECT_INPUT);
        }
        if (integer > LottoRules.LAST_NUM) {
            throw new RuntimeException(INCORRECT_INPUT);
        }
```

File: src/main/java/lotto/OutputView.java [hunk 16]
```diff
@@ -0,0 +1,63 @@
+package lotto;
```
Comment (hyunniiii) [inline src/main/java/lotto/OutputView.java:1]: View, Controller, Domain 패키지 생성 후 각 패키지 하위로 넣어주세요 ″̮

**Review by hyunniiii — CHANGES_REQUESTED** [review]
안녕하세요 민상님~
몇가지 코멘트 남겼으니 확인 부탁드립니다 ″̮

File: src/main/java/calculator/InputView.java [hunk 17]
```diff
@@ -0,0 +1,20 @@
...
+        String userInput = scanner.nextLine();
+        return checkInputNullOrZero(userInput);
+    }
+    
+    private static String checkInputNullOrZero(String userInput) {
+        if (userInput == null || userInput.equals("")) {
+            return "0";
+        }
+        return userInput;
+    }
```
[the remark this answers is not in the document]
Comment (minsang0850) [reply src/main/java/calculator/InputView.java:19]: private 메소드는 어떻게 테스트 해야 하나요?
checkInputNullOrZero를 호출하는 inputExpression도 입력을 받는 메소드라 어떻게 테스트해야하는지 궁금합니다!

File: src/main/java/lotto/InputView.java [hunk 18]
```diff
@@ -0,0 +1,51 @@
+package lotto;
+
+import java.util.List;
+import java.util.Scanner;
+
+public class InputView {
+
+    private static final Scanner sc = new Scanner(System.in);
```
[the remark this answers is not in the document]
Comment (minsang0850) [reply src/main/java/lotto/InputView.java:8]: 넵

File: src/main/java/calculator/Operand.java [hunk 19]
```diff
@@ -0,0 +1,49 @@
...
+            return Integer.parseInt(parseResult);
+        } catch (NumberFormatException e) {
+            throw new NumberFormatException("피연산자가 숫자가 아닙니다.");
+        }
+        
+    }
+    
+    private void checkNegative(int operand) {
+        if (operand < 0) {
+            throw new RuntimeException("음수입니다");
```
[the remark this answers is not in the document]
Comment (minsang0850) [reply src/main/java/calculator/Operand.java:29]: 넵

File: src/main/java/lotto/Statistics.java (hunk 14 above)
Comment (minsang0850) [reply src/main/java/lotto/Statistics.java:29]: 무슨 말씀이신지 더 간단하게 설명해주실 수 있나요?

File: src/main/java/calculator/InputView.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (hyunniiii) [reply src/main/java/calculator/InputView.java:19]: 테스트하기가 어렵다면 테스트 하는 위치를 옮겨도 좋을 것 같아요. ″̮ 
view단에서 테스트를 하기 힘들다면 객체를 이용해서 테스트해보는건 어떨까요?

File: src/main/java/calculator/Operand.java [hunk 20]
```diff
@@ -0,0 +1,49 @@
...
+        return operand;
+    }
+    
+    private int parseStringToInteger(String parseResult) {
+        try {
+            return Integer.parseInt(parseResult);
+        } catch (NumberFormatException e) {
+            throw new NumberFormatException("피연산자가 숫자가 아닙니다.");
+        }
+
```
Comment (hyunniiii) [inline src/main/java/calculator/Operand.java:24]: 불필요한 개행은 제거해주세요 ″̮

File: src/main/java/lotto/LottoAuto.java [hunk 21]
```diff
@@ -0,0 +1,17 @@
...
+public class LottoAuto {
+    
+    public void play() {
+        LottoRules lottoRules = new LottoRules();
+        int lottoCount = InputView.getPrice() / lottoRules.LOTTO_PRICE;
+        User user = new User(
+          OutputView.getLottoCount(lottoCount), lottoRules);
+        OutputView.printLottos(user.getLottos());
+        WinningNumber winningNumber = new WinningNumber(InputView.getWinningNumber(), lottoRules);
+        winningNumber.setBonusBall(InputView.getBonusBall());
```
Comment (hyunniiii) [inline src/main/java/lotto/LottoAuto.java]: `setBonusBall`를 해주기보다는 객체 생성할 때 생성인자로 받아서 winningNumber와 함께 한번에 초기화해주는건 어떨까요?

File: src/main/java/lotto/Statistics.java (hunk 14 above)
Comment (hyunniiii) [reply src/main/java/lotto/Statistics.java]: setRanks를 통해서 외부에서 Map<Rank,Integer> ranks의 값을 변경해주고 있어요.
이렇게 다른 곳에서 값을 변경하게 된다면 결과가 항상 바뀌게 됩니다. 

즉, `Map<Rank, Integer> ranks`를 한번만 초기화 하여 불변을 유지하도록 해주세요 ″̮ 

https://jinwooe.wordpress.com/2017/12/21/%EB%B6%80%EC%88%98-%ED%9A%A8%EA%B3%BC-side-effect-%EC%B0%B8%EC%A1%B0-%ED%88%AC%EB%AA%85%EC%84%B1-referential-transparency/

**Review by hyunniiii — APPROVED** [review]
안녕하세요 민상님~! 
피드백 반영 잘해주셨네요 ″̮ 
몇가지 코멘트 남겼으니 다음 미션에 함께 반영 부탁드립니다!
